File: bindings/javascript/tests/conformance.test.js
'use strict';

/**
 * Cross-language conformance suite.
 *
 * Reads tests/conformance/cases.txt — the corpus every implementation shares —
 * and checks this binding against the C-generated fixtures: decoding a fixture
 * must produce the corpus values, and encoding the corpus values must produce
 * the fixture byte for byte.
 *
 * See tests/conformance/README.md.
 */

const fs = require('fs');
const path = require('path');
const { encode, decode } = require('../src/index');

const CONFORMANCE_DIR = path.resolve(__dirname, '../../../tests/conformance');
const CASES_FILE = path.join(CONFORMANCE_DIR, 'cases.txt');
const FIXTURE_DIR = path.join(CONFORMANCE_DIR, 'fixtures');
const MALFORMED_DIR = path.join(CONFORMANCE_DIR, 'malformed');

// ── Corpus parsing ───────────────────────────────────────────────────────

/** Decode a corpus token ("~" for empty, %XX escapes) into a Buffer. */
function tokenBytes(tok) {
    if (tok === '~') return Buffer.alloc(0);
    const out = [];
    for (let i = 0; i < tok.length;) {
        if (tok[i] === '%') {
            out.push(parseInt(tok.substr(i + 1, 2), 16));
            i += 3;
        } else {
            out.push(tok.charCodeAt(i));
            i += 1;
        }
    }
    return Buffer.from(out);
}

function tokenString(tok) {
    return tokenBytes(tok).toString('utf8');
}

function hexBytes(tok) {
    if (tok === '~') return Buffer.alloc(0);
    return Buffer.from(tok, 'hex');
}

function f64FromHex(tok) {
    const b = Buffer.from(tok, 'hex');
    return b.readDoubleBE(0);
}

function f32FromHex(tok) {
    const b = Buffer.from(tok, 'hex');
    return b.readFloatBE(0);
}

function parseCases(text) {
    const cases = [];
    let cur = null;
    for (const raw of text.split('\n')) {
        const line = raw.trim();
        if (!line || line.startsWith('#')) continue;
        const parts = line.split(' ');
        if (parts[0] === 'case') {
            cur = {
                name: parts[1],
                encodeExpected: !parts.includes('encode=no'),
                // JavaScript holds integers in a double, so cases that need
                // the full 64-bit range are not this binding's to check.
                skip: parts.includes('requires=int64'),
                keys: [],
            };
            cases.push(cur);
        } else if (parts[0] === 'key') {
            cur.keys.push({ key: tokenString(parts[1]), type: parts[2], arg: parts[3] });
        } else {
            throw new Error('unknown corpus directive: ' + parts[0]);
        }
    }
    return cases;
}

/** Build the JS value a corpus entry describes. */
function caseValue(entry) {
    switch (entry.type) {
        case 'null': return null;
        case 'bool': return entry.arg === '1';
        case 'int': return Number(entry.arg);
        case 'uint': return Number(entry.arg);
        case 'f64': return f64FromHex(entry.arg);
        case 'f32': return f32FromHex(entry.arg);
        case 'str': return tokenString(entry.arg);
        case 'blob': return hexBytes(entry.arg);
        default: throw new Error('unknown corpus type: ' + entry.type);
    }
}

function caseData(c) {
    const data = {};
    for (const entry of c.keys) data[entry.key] = caseValue(entry);
    return data;
}

/**
 * Compare a decoded value with the expected one. Doubles compare by bit
 * pattern so -0.0 and NaN are handled, with all NaNs treated as equal.
 */
function valuesMatch(got, want) {
    if (want === null) return got === null;
    if (typeof want === 'boolean') return got === want;
    if (typeof want === 'number') {
        if (typeof got !== 'number') return false;
        if (Number.isNaN(want) && Number.isNaN(got)) return true;
        return Object.is(got, want);
    }
    if (typeof want === 'string') return got === want;
    if (want instanceof Uint8Array) {
        return got instanceof Uint8Array && Buffer.from(got).equals(Buffer.from(want));
    }
    return false;
}

// ── The suite ────────────────────────────────────────────────────────────

const cases = parseCases(fs.readFileSync(CASES_FILE, 'utf8'));

describe('cross-language conformance corpus', () => {
    test('corpus and fixtures are present', () => {
        expect(cases.length).toBeGreaterThan(0);
        for (const c of cases) {
            expect(fs.existsSync(path.join(FIXTURE_DIR, c.name + '.trp'))).toBe(true);
        }
    });

    describe.each(cases.map(c => [c.name, c]))('%s', (_name, c) => {
        const fixture = () => new Uint8Array(fs.readFileSync(path.join(FIXTURE_DIR, c.name + '.trp')));

        const decodeTest = c.skip ? test.skip : test;
        decodeTest('decodes the C fixture to the corpus values', () => {
            const result = decode(fixture());
            const want = caseData(c);
            expect(Object.keys(result).sort()).toEqual(Object.keys(want).sort());
            for (const k of Object.keys(want)) {
                expect(valuesMatch(result[k], want[k])).toBe(true);
            }
        });

        // Cases marked encode=no describe values this binding cannot
        // reproduce exactly (float32, or doubles JS reads back as integers).
        const encodeTest = c.encodeExpected && !c.skip ? test : test.skip;
        encodeTest('encodes byte-for-byte like the C reference', () => {
            const got = Buffer.from(encode(caseData(c)));
            expect(got.equals(Buffer.from(fixture()))).toBe(true);
        });
    });
});

// Every buffer here is a valid fixture with one field damaged. Damage inside
// the data is re-sealed with a correct CRC, so the reader has to catch it
// rather than being handed a checksum failure.
describe('malformed inputs are rejected', () => {
    const files = fs.readdirSync(MALFORMED_DIR).filter(f => f.endsWith('.trp')).sort();

    test('the malformed corpus is present', () => {
        expect(files.length).toBeGreaterThan(0);
    });

    test.each(files)('%s', name => {
        const buf = new Uint8Array(fs.readFileSync(path.join(MALFORMED_DIR, name)));
        expect(() => decode(buf)).toThrow();
    });
});

File: bindings/rust/tests/conformance_test.rs
// Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.

//! Cross-language conformance suite.
//!
//! Reads `tests/conformance/cases.txt` -- the corpus every implementation
//! shares -- and checks this binding against the C-generated fixtures: decoding
//! a fixture must produce the corpus values, and encoding the corpus values
//! must produce the fixture byte for byte.
//!
//! See `tests/conformance/README.md`.

use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

use triepack::{decode, encode, Value};

fn conformance_dir() -> PathBuf {
    let mut p = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    p.push("..");
    p.push("..");
    p.push("tests");
    p.push("conformance");
    p
}

// -- Corpus parsing --------------------------------------------------------

/// Decode a corpus token ("~" for empty, %XX escapes) into bytes.
fn token_bytes(tok: &str) -> Vec<u8> {
    if tok == "~" {
        return Vec::new();
    }
    let b = tok.as_bytes();
    let mut out = Vec::with_capacity(b.len());
    let mut i = 0;
    while i < b.len() {
        if b[i] == b'%' {
            let hi = (b[i + 1] as char).to_digit(16).expect("bad escape");
            let lo = (b[i + 2] as char).to_digit(16).expect("bad escape");
            out.push((hi * 16 + lo) as u8);
            i += 3;
        } else {
            out.push(b[i]);
            i += 1;
        }
    }
    out
}

fn token_string(tok: &str) -> String {
    String::from_utf8(token_bytes(tok)).expect("corpus token is not UTF-8")
}

fn hex_bytes(tok: &str) -> Vec<u8> {
    if tok == "~" {
        return Vec::new();
    }
    (0..tok.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&tok[i..i + 2], 16).expect("bad hex"))
        .collect()
}

struct Case {
    name: String,
    encode_expected: bool,
    entries: Vec<(String, String, Option<String>)>,
}

impl Case {
    fn fixture(&self) -> Vec<u8> {
        let mut p = conformance_dir();
        p.push("fixtures");
        p.push(format!("{}.trp", self.name));
        fs::read(&p).unwrap_or_else(|e| panic!("case {}: {}: {}", self.name, p.display(), e))
    }

    fn data(&self) -> HashMap<String, Value> {
        self.entries
            .iter()
            .map(|(k, t, arg)| (k.clone(), corpus_value(t, arg.as_deref())))
            .collect()
    }
}

/// Build the Rust value a corpus entry describes.
fn corpus_value(type_name: &str, arg: Option<&str>) -> Value {
    match type_name {
        "null" => Value::Null,
        "bool" => Value::Bool(arg == Some("1")),
        "int" => Value::Int(arg.unwrap().parse().expect("bad int")),
        "uint" => Value::UInt(arg.unwrap().parse().expect("bad uint")),
        "f64" => Value::Float64(f64::from_bits(
            u64::from_str_radix(arg.unwrap(), 16).expect("bad f64"),
        )),
        "f32" => Value::Float64(
            f32::from_bits(u32::from_str_radix(arg.unwrap(), 16).expect("bad f32")) as f64,
        ),
        "str" => Value::String(token_string(arg.unwrap())),
        "blob" => Value::Blob(hex_bytes(arg.unwrap())),
        other => panic!("unknown corpus type: {}", other),
    }
}

fn parse_cases() -> Vec<Case> {
    let mut p = conformance_dir();
    p.push("cases.txt");
    let text = fs::read_to_string(&p).unwrap_or_else(|e| panic!("{}: {}", p.display(), e));

    let mut cases: Vec<Case> = Vec::new();
    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let parts: Vec<&str> = line.split(' ').collect();
        match parts[0] {
            "case" => cases.push(Case {
                name: parts[1].to_string(),
                encode_expected: !parts.contains(&"encode=no"),
                entries: Vec::new(),
            }),
            "key" => {
                let cur = cases.last_mut().expect("key outside a case");
                cur.entries.push((
                    token_string(parts[1]),
                    parts[2].to_string(),
                    parts.get(3).map(|s| s.to_string()),
                ));
            }
            other => panic!("unknown corpus directive: {}", other),
        }
    }
    cases
}

/// Compare a decoded value with the expected one. Doubles compare by bit
/// pattern so -0.0 is distinguished, with all NaNs treated as equal.
fn values_match(got: &Value, want: &Value) -> bool {
    match (got, want) {
        (Value::Float64(g), Value::Float64(w)) => {
            (g.is_nan() && w.is_nan()) || g.to_bits() == w.to_bits()
        }
        _ => got == want,
    }
}

// -- The suite -------------------------------------------------------------

#[test]
fn conformance_corpus_and_fixtures_present() {
    let cases = parse_cases();
    assert!(!cases.is_empty(), "corpus is empty");
    for c in &cases {
        let mut p = conformance_dir();
        p.push("fixtures");
        p.push(format!("{}.trp", c.name));
        assert!(p.exists(), "case {}: missing fixture", c.name);
    }
}

#[test]
fn conformance_decodes_c_fixtures() {
    for c in parse_cases() {
        let result = decode(&c.fixture())
            .unwrap_or_else(|e| panic!("case {}: decode failed: {}", c.name, e));
        let want = c.data();
        assert_eq!(result.len(), want.len(), "case {}: key count", c.name);
        for (k, v) in &want {
            let got = result
                .get(k)
                .unwrap_or_else(|| panic!("case {}: missing key {:?}", c.name, k));
            assert!(
                values_match(got, v),
                "case {}: key {:?}: got {:?}, want {:?}",
                c.name,
                k,
                got,
                v
            );
        }
    }
}

#[test]
fn conformance_encodes_like_c_reference() {
    for c in parse_cases() {
        // Cases marked encode=no describe values this binding cannot
        // reproduce exactly (float32, or doubles some languages read back as
        // integers).
        if !c.encode_expected {
            continue;
        }
        let got = encode(&c.data());
        let want = c.fixture();
        assert_eq!(
            got,
            want,
            "case {}: encoded {} bytes, C reference is {}",
            c.name,
            got.len(),
            want.len()
        );
    }
}

// -- Malformed inputs ------------------------------------------------------

/// Each file is a valid fixture with one field damaged. Damage inside the data
/// is re-sealed with a correct CRC, so the reader has to catch it rather than
/// being handed a checksum failure.
#[test]
fn conformance_rejects_malformed() {
    let mut dir = conformance_dir();
    dir.push("malformed");
    let mut seen = 0;
    for entry in fs::read_dir(&dir).expect("cannot read malformed corpus") {
        let path = entry.expect("bad dir entry").path();
        if path.extension().and_then(|e| e.to_str()) != Some("trp") {
            continue;
        }
        seen += 1;
        let buf = fs::read(&path).expect("cannot read fixture");
        assert!(
            decode(&buf).is_err(),
            "{}: accepted malformed input",
            path.display()
        );
    }
    assert!(seen > 0, "malformed corpus is empty");
}
